### PR TITLE
Kroach/oasis 1567 thirdparty

### PR DESCRIPTION
**Pull request by kellyroach-optimizely**
PULL REQUEST PURPOSE

Currently, Optimizely objective-c-sdk depends on third party repos
    *FMDB https://github.com/ccgus/fmdb
    *JSONModel https://github.com/jsonmodel/jsonmodel
The ostensible benefits of doing so seem negated lately by the price incurred
by customers who wish to link their own customized FMDB or JSONModel
or by our strange Carthage Cartfile tied to a specific JSONModel commit
"1e80ecaec8314c7d367676ed0e8545c10c563612" .

This pull request proposes to resolve these issues by rolling renamed
FMDB --> OPDB and JSONModel --> OPJMModel into OptimizelySDKShared
and OptimizelySDKCore .  The pods FMDB and JSONModel are demoted
from pods to included source code, still given proper attributions with their
included licenses and copyrights.  Modifications Optimizely makes are Apache 2 
and so noted in our own added copyrights.  We've also added test suites
OPDBTests and OPJMModelTests .

### CODE REVIEW

**Comment by kellyroach-optimizely** [comment]
TEST:
* 8 iOS test suites PASSED
* OptimizelySDKiOSUniversalTests PASSED
* OptimizelyiOSDemoApp PASSED
* 7 tvOS test suites PASSED
* OptimizelySDKTVOSUniversalTests PASSED
* OptimizelyTVOSDemoApp PASSED
* OptimizelyiOSDemoApp / "Generic iOS Device" / Team = "Kelly Roach" / Product > Archive PASSED
Validate... is still FLAKEY (unsure)
* OptimizelySDKiOS Product > Analyze
Build succeeded / No issues
PASSED
* Above includes OPDBTests and OPJMModeliOSTests

**Review by alda-optimizely — DISMISSED** [review]
• Can we put `OPJMModel` and `OPDB` in the Frameworks folder? 

• I still have a preference for `OPTLY` prefix to be consistent with our SDK files:
`OPTLYFMDB`
`OPTLYJSONModel`
It would also be nice when we try to search for our code/files in a stack trace — we wouldn’t need to search through several prefixes. 
But, if this is a big change, then you can leave it. 

-Podfile:
Remove all the entries with non test targets:
`target 'OptimizelySDKiOS' do
  project 'OptimizelySDKiOS/OptimizelySDKiOS.xcodeproj/'
  platform :ios, '8.0'
end
`

• In the non tests **Build Phases**, we should clean up all the CocoaPods files. For example, remove:
[CP] Check Pods Manifest.lock
[CP] Copy Pods Resources
All *xcconfig files in the Pods folder

**Review by wangjoshuah — CHANGES_REQUESTED** [review]
@kellyroach-optimizely the last two builds failed on travis. can you try and fix whatever is causing the issue?

File: Scripts/build_all.sh [hunk 1]
```diff
@@ -49,34 +49,22 @@ main() {
...
-  xcodebuild -workspace OptimizelySDK.xcworkspace -scheme Pods-OptimizelySDKiOS -configuration Release "${action}"
   xcodebuild -workspace OptimizelySDK.xcworkspace -scheme Pods-OptimizelySDKiOSTests -configuration Release "${action}"
-  xcodebuild -workspace OptimizelySDK.xcworkspace -scheme Pods-OptimizelySDKSharediOS -configuration Release "${action}"
   xcodebuild -workspace OptimizelySDK.xcworkspace -scheme Pods-OptimizelySDKSharediOSTests -configuration Release "${action}"
-  xcodebuild -workspace OptimizelySDK.xcworkspace -scheme Pods-OptimizelySDKSharedTVOS -configuration Release "${action}"
   xcodebuild -workspace OptimizelySDK.xcworkspace -scheme Pods-OptimizelySDKSharedTVOSTests -configuration Release "${action}"
-  xcodebuild -workspace OptimizelySDK.xcworkspace -scheme Pods-OptimizelySDKTVOS -configuration Release "${action}"
   xcodebuild -workspace OptimizelySDK.xcworkspace -scheme Pods-OptimizelySDKTVOSTests -configuration Release "${action}"
-  xcodebuild -workspace OptimizelySDK.xcworkspace -scheme Pods-OptimizelySDKUserProfileServiceiOS -configuration Release "${action}"
```
Comment (wangjoshuah) [inline Scripts/build_all.sh:77]: why did we remove these build commands from the build-all script?

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYEventMetric.h [hunk 2]
```diff
@@ -18,9 +18,9 @@
 
 #import <Foundation/Foundation.h>
 #ifdef UNIVERSAL
-    #import "JSONModelLib.h"
+    #import "OPTLYJSONModelLib.h"
```
Comment (wangjoshuah) [inline OptimizelySDKCore/OptimizelySDKCore/OPTLYEventMetric.h:21]: can this all be moved to a prefix header?

**Review by wangjoshuah — CHANGES_REQUESTED** [review]

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYEventMetric.h (hunk 2 above)
Comment (wangjoshuah) [reply OptimizelySDKCore/OptimizelySDKCore/OPTLYEventMetric.h:21]: that way the next time we need a refactor like this, we will only need to change the prefix header

File: OptimizelySDKUniversal/unexported_symbols.txt [hunk 3]
```diff
@@ -1,73 +1,15 @@
...
-_OPDBBlockSQLiteCallBackFunction
-_OPDBExecuteBulkSQLCallback
 _HASH_SEED
-_OPJMModelErrorDomain
 _MAX_HASH_VALUE
 _MAX_TRAFFIC_VALUE
 _MurmurHash3_x64_128
 _MurmurHash3_x86_128
 _MurmurHash3_x86_32
-_OBJC_CLASS_$_OPDBDatabase
```
Comment (wangjoshuah) [inline OptimizelySDKUniversal/unexported_symbols.txt]: if someone wanted to use FMDB and our universal framework. would this prevent them from using our modified FMDB?

File: OptimizelySDKCore/OptimizelySDKCore.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -192,8 +192,8 @@
 		EA16D93D1ECBD90E00C4C998 /* OPTLYExperimentBucketMapEntity.h in Headers */ = {isa = PBXBuildFile; fileRef = EA16D93A1ECBD90E00C4C998 /* OPTLYExperimentBucketMapEntity.h */; settings = {ATTRIBUTES = (Public, ); }; };
 		EA16D93E1ECBD90E00C4C998 /* OPTLYExperimentBucketMapEntity.m in Sources */ = {isa = PBXBuildFile; fileRef = EA16D93B1ECBD90E00C4C998 /* OPTLYExperimentBucketMapEntity.m */; };
 		EA16D93F1ECBD90E00C4C998 /* OPTLYExperimentBucketMapEntity.m in Sources */ = {isa = PBXBuildFile; fileRef = EA16D93B1ECBD90E00C4C998 /* OPTLYExperimentBucketMapEntity.m */; };
-		EA16D9421ECCC5C600C4C998 /* OPTLYDecisionService.h in Headers */ = {isa = PBXBuildFile; fileRef = EA16D9401ECCC5C600C4C998 /* OPTLYDecisionService.h */; };
-		EA16D9431ECCC5C600C4C998 /* OPTLYDecisionService.h in Headers */ = {isa = PBXBuildFile; fileRef = EA16D9401ECCC5C600C4C998 /* OPTLYDecisionService.h */; };
+		EA16D9421ECCC5C600C4C998 /* OPTLYDecisionService.h in Headers */ = {isa = PBXBuildFile; fileRef = EA16D9401ECCC5C600C4C998 /* OPTLYDecisionService.h */; settings = {ATTRIBUTES = (Public, ); }; };
```
Comment (wangjoshuah) [inline OptimizelySDKCore/OptimizelySDKCore.xcodeproj/project.pbxproj:193]: why does this need to be public?

File: OptimizelySDKUniversal/OptimizelySDKUniversal.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -37,10 +37,10 @@
 		EA3144E91ED7A1C100A8E555 /* OPTLYDecisionService.m in Sources */ = {isa = PBXBuildFile; fileRef = EA3144E41ED7A19700A8E555 /* OPTLYDecisionService.m */; };
 		EA3144EA1ED7A1C100A8E555 /* OPTLYExperimentBucketMapEntity.m in Sources */ = {isa = PBXBuildFile; fileRef = EA3144E61ED7A19700A8E555 /* OPTLYExperimentBucketMapEntity.m */; };
 		EA3144EB1ED7A1C100A8E555 /* OPTLYUserProfile.m in Sources */ = {isa = PBXBuildFile; fileRef = EA3144E81ED7A19700A8E555 /* OPTLYUserProfile.m */; };
-		EA3144EC1ED7A1CD00A8E555 /* OPTLYDecisionService.h in Headers */ = {isa = PBXBuildFile; fileRef = EA3144E31ED7A19700A8E555 /* OPTLYDecisionService.h */; };
+		EA3144EC1ED7A1CD00A8E555 /* OPTLYDecisionService.h in Headers */ = {isa = PBXBuildFile; fileRef = EA3144E31ED7A19700A8E555 /* OPTLYDecisionService.h */; settings = {ATTRIBUTES = (Public, ); }; };
```
Comment (wangjoshuah) [inline OptimizelySDKUniversal/OptimizelySDKUniversal.xcodeproj/project.pbxproj:40]: again, why does this need to be public?

File: Podfile [hunk 6]
```diff
@@ -1,7 +1,7 @@
 workspace 'OptimizelySDK.xcworkspace'
 
 def common_pods
-    pod 'JSONModel', '1.3.0'
+    pod 'JSONModel', :git => 'https://github.com/jsonmodel/jsonmodel.git', :commit => '364f57b'
```
Comment (wangjoshuah) [inline Podfile]: aren't we not using JSONModel and FMDB anymore?

**Review by wangjoshuah — CHANGES_REQUESTED** [review]

File: OptimizelySDKCore/OptimizelySDKCore.xcodeproj/project.pbxproj [hunk 7]
```diff
@@ -2054,7 +2054,7 @@
 					"$(inherited)",
 				);
 				INFOPLIST_FILE = "OPTLYJSONModelTests/OPTLYJSONModelTests-Info.plist";
-				IPHONEOS_DEPLOYMENT_TARGET = 10.3;
+				IPHONEOS_DEPLOYMENT_TARGET = 8.0;
```
Comment (wangjoshuah) [inline OptimizelySDKCore/OptimizelySDKCore.xcodeproj/project.pbxproj]: we don't need to test through iOS 8, we have minimum compatibility with iOS 9.

**Review by wangjoshuah — CHANGES_REQUESTED** [review]

**Review by wangjoshuah — APPROVED** [review]

**Comment by wangjoshuah** [comment]
everything seems to be working fine when you land this on master can you be sure to run the e2e tests for iOS to make sure nothing breaks?

**Comment by kellyroach-optimizely** [comment]
RESPONSES TO ALDA LUONG'S COMMENTS

COMMENT#1: Can we put OPJMModel and OPDB in the Frameworks folder?
RESPONSE: JSONModel and FMDB files we keep are absorbed into
2 existing Optimizely frameworks Core and Shared .  The PRO's of
this approach: No increase in number of distributed Optimizely
CocoaPods AND considerable Optimizely SDK size reduction (16% or 11%
depending on what you count) due to in part o fewer framework overheads.

COMMENT#2: preference for OPTLY prefix ... OPTLYFMDB OPTLYJSONModel
 @kellyroach-optimizely	OPDB --> OPTLYFMDB	Verified		fd32307
 @kellyroach-optimizely	OPJM --> OPTLYJSON	Verified		ced54db
DONE

COMMENT#3: Podfile Remove all the entries with non test targets
 @kellyroach-optimizely	Remove Podfile targets which no longer depend on pods	Verified		6b5896e
 @kellyroach-optimizely	pod install	Verified		fa26a68
 @kellyroach-optimizely	Update build_all.sh	Verified		ff2c39c
DONE

COMMENT#4: non tests Build Phases clean up all the CocoaPods [CP] ...
Same as for COMMENT#3
DONE

File: OptimizelySDKCore/OptimizelySDKCore.xcodeproj/project.pbxproj (hunk 4 above)
Comment (kellyroach-optimizely) [reply OptimizelySDKCore/OptimizelySDKCore.xcodeproj/project.pbxproj:193]: Here is the compiler warning a customer would
see following the README.md "Manual Instructions" if we did
not make OPTLYDecisionService.h be 'Public':

<module-includes>:1:1: warning: umbrella header for module 'OptimizelySDKiOS' does not include header 'OPTLYDecisionService.h' [-Wincomplete-umbrella]
#import "Headers/OptimizelySDKiOS.h"
^
1 warning generated.
1 warning generated.

File: OptimizelySDKUniversal/OptimizelySDKUniversal.xcodeproj/project.pbxproj (hunk 5 above)
Comment (kellyroach-optimizely) [reply OptimizelySDKUniversal/OptimizelySDKUniversal.xcodeproj/project.pbxproj:40]: Same answer to the same question.

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYEventMetric.h (hunk 2 above)
Comment (kellyroach-optimizely) [reply OptimizelySDKCore/OptimizelySDKCore/OPTLYEventMetric.h:21]: Possible CON's to the prefix header proposal:

* Apple and Apple fans are discouraging *.pch's these days.
"Why precompiled header dropped in recent Xcode versions?
... Correct answer by Justin [410 points - Apple Staff]
Precompiled Header Files are discouraged in favor of modules. ..."
https://forums.developer.apple.com/thread/22060
"4 Ways Precompiled Headers Cripple Your Code"
https://qualitycoding.org/precompiled-header/

* Typically, the single *.pch can't be made 'Public', but we are creating a *.framework being distributed to the public.  Multiple *.h can be sorted into 'Public' and 'Private'.

* Knowing this P.R. would already involve a lot of change and commits, I've been
seeking to not bite off more than what's required to achieve the PULL REQUEST PURPOSE. Creating *.pch is unnecessary for the PURPOSE but the rename is already necessary and suffices for this line of code.

File: Scripts/build_all.sh (hunk 1 above)
Comment (kellyroach-optimizely) [reply Scripts/build_all.sh:77]: "pod install" and "pod update" automatically calculate which Pods-Xxx are added to the workspace.  We've done work which affects that calculation, and with fewer autogenerated Pods-Xxx coming from "pod", our "build_all.sh" can be smaller.

Many of the "non tests" Xxx targets no longer depend on any pods and previously may have depended on FMDB and/or JSONModel . Those Xxx targets which no longer depend on pods have their Pods-Xxx go away since these CocoaPods automatically generated targets no longer exist.

File: OptimizelySDKCore/OptimizelySDKCore.xcodeproj/project.pbxproj (hunk 7 above)
Comment (kellyroach-optimizely) [reply OptimizelySDKCore/OptimizelySDKCore.xcodeproj/project.pbxproj]: I've taken your advice and have changed local_travis.sh to only test >= 9.0 which saves us from cluttering up the OPTLYJSONModelTests with any new code which is sensitive to the iOS version.

@kellyroach-optimizely	Change 8.4 to 9.0 in local_travis.sh	Verified		ba69951

**Comment by kellyroach-optimizely** [comment]
RESPONSES TO JOSH WANG'S COMMENTS

COMMENT#1: @kellyroach-optimizely the last two builds failed on travis.
RESPONSE: Improved to the point that Travis DID pass.
@kellyroach-optimizely	Change 'Deployment Target' of 3 new tests to iOS 8.0 / tvOS 9.0	Verified		09fa48e
@kellyroach-optimizely	Change 8.4 to 9.0 in local_travis.sh	Verified		ba69951
@kellyroach-optimizely	Update README.md with use_frameworks	Verified		f5e69d5
Did a LOT of analysis today of which pairs of (OS,name) simulators in the local_travis.sh xcodebuilds are passing and the few which aren't.  We're also checking if the OptimizelyiOSDemoApp is nevertheless behaving fine.

COMMENT#2: can you be sure to run the e2e tests for iOS to make sure nothing breaks?
RESPONSE: Jenkins "OASIS" tab is currently showing [green checkmark]'s on our "iOS..." and "Obj-C..." E2E and other tests.

File: OptimizelySDKCore/OptimizelySDKCore.xcodeproj/project.pbxproj (hunk 4 above)
Comment (wangjoshuah) [reply OptimizelySDKCore/OptimizelySDKCore.xcodeproj/project.pbxproj:193]: does it need to be in the umbrella? The customer never personally interacts with it right?

**Comment by kellyroach-optimizely** [comment]
TEST: (After 'iPad Retina' --> 'iPad Air' in local_travis.sh	Verified		6efdc6b)
* Remote Travis PASSED
* sh ./Scripts/local_travis.sh PASSED
* OptimizelyiOSDemoApp PASSED
* 3 most critical iOS test suites PASSED
* OptimizelySDKiOS Product > Analyze PASSED
* OptimizelyiOSDemoApp Product > Analyze PASSED
* OptimizelyiOSDemoApp Product > Archive PASSED
* OptimizelyTVOSDemoApp PASSED
* OptimizelyTVOSDemoApp Product > Analyze PASSED
* Clean / build both 2 universal targets PASSED
* sh ./Scripts/build_all.sh PASSED

CONCLUDE: PASSED
GREEN LIGHT for P.R.#170 "Squash merge".
